tests/infer/test_enum.py

```diff
@@ -1514,29 +1524,34 @@ def hand_model(data):
         probs_x = pyro.param("model_probs_x")
         probs_y = pyro.param("model_probs_y")
         probs_z = pyro.param("model_probs_z")
-        probs_yz = probs_y.mm(probs_z)
         x = pyro.sample("x", dist.Categorical(probs_x))
-        with pyro.iarange("data", num_masked):
-            with poutine.scale(scale=scale):
-                pyro.sample("z", dist.Categorical(probs_yz[x]), obs=data[:num_masked])
+        with poutine.scale(scale=scale):
+            for i in pyro.irange("data", num_masked):
+                y = pyro.sample("y_{}".format(i), dist.Categorical(probs_y[x]),
+                                infer={"enumerate": "parallel"})
+                pyro.sample("z_{}".format(i), dist.Categorical(probs_z[y]), obs=data[i])
 
     @config_enumerate(default="parallel")
     def guide(data):
         probs_x = pyro.param("guide_probs_x")
         pyro.sample("x", dist.Categorical(probs_x))
 
     data = dist.Categorical(torch.tensor([0.3, 0.7])).sample((num_samples,))
-    elbo = TraceEnum_ELBO(max_iarange_nesting=1, strict_enumeration_warning=False)
+    elbo = TraceEnum_ELBO(max_iarange_nesting=1)
     auto_loss = elbo.differentiable_loss(auto_model, guide, data)
     hand_loss = elbo.differentiable_loss(hand_model, guide, data)
     _check_loss_and_grads(hand_loss, auto_loss)
 
 
 @pytest.mark.parametrize('scale', [1, 10])
 @pytest.mark.parametrize('num_samples,num_masked',
-                         [(1, 1), (3, 3), (3, 1)],
+                         [(1, 1), (2, 2), (3, 2)],
                          ids=["single", "batch", "masked"])
 def test_elbo_enumerate_iarange_3(num_samples, num_masked, scale):
+    #  +-----------------------+
+    #  | x ----> y ----> z     |
+    #  |                     N |
+    #  +-----------------------+
```

Comment:
I don't think of these tests as not so much a list of specific regression tests exercising edge cases. Rather I think of these as an exhaustive enumeration of models up to a given complexity bound.

I don't think it would be helpful to add text describing the history of the kind of bug each test originally caught.

Reply:
Ideally we'd automatically generate these tests by enumerating ASTs up to a given complexity. In this PR I did that manually. The first set of tests is all iarange nestings of 3 variables and 1 iarange; the next set of tests should be all iarange nestings with two iaranges and at most 4 variables (with duplicates removed).

Reply:
That would be neat, but it sounds like it could get complex enough that we'd need tests for those tests

Reply:
Ok I've added some comments.